integrations/tensorflow/e2e/iree_e2e_cartesian_product_test_suite.cmake

```diff
@@ -0,0 +1,247 @@
+#             python range_test.py  --target_backend=iree_llvmaot --reference_backend=tf --magic_flag=true
+#   DISABLED: python range_test.py  --target_backend=iree_vulkan  --reference_backend=tf --magic_flag=true
+#             python concat_test.py --target_backend=tf           --reference_backend=tf --magic_flag=false
+#             python concat_test.py --target_backend=tflite       --reference_backend=tf --magic_flag=false
+#             python concat_test.py --target_backend=iree_vmla    --reference_backend=tf --magic_flag=false
+#             python concat_test.py --target_backend=iree_llvmaot --reference_backend=tf --magic_flag=false
+#             python concat_test.py --target_backend=iree_vulkan  --reference_backend=tf --magic_flag=false
+#             python range_test.py  --target_backend=tf           --reference_backend=tf --magic_flag=false
+#             python range_test.py  --target_backend=tflite       --reference_backend=tf --magic_flag=false
+#             python range_test.py  --target_backend=iree_vmla    --reference_backend=tf --magic_flag=false
+#             python range_test.py  --target_backend=iree_llvmaot --reference_backend=tf --magic_flag=false
+#   DISABLED: python range_test.py  --target_backend=iree_vulkan  --reference_backend=tf --magic_flag=false
+#
+function(iree_e2e_cartesian_product_test_suite)
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    PARSE_ARGV 0
+    _RULE
+    ""
+    "NAME"
+    "MATRIX_KEYS;MATRIX_VALUES;FAILING_CONFIGURATIONS;LABELS"
+  )
+
+  list(LENGTH _RULE_MATRIX_KEYS _MATRIX_KEYS_COUNT)
+  list(LENGTH _RULE_MATRIX_VALUES _MATRIX_VALUES_COUNT)
+
+  if(NOT _MATRIX_KEYS_COUNT EQUAL _MATRIX_VALUES_COUNT)
+    message(
+      SEND_ERROR
+        "MATRIX_KEYS count ${_MATRIX_KEYS_COUNT} does not match MATRIX_VALUES count ${_MATRIX_VALUES_COUNT}"
+    )
+  endif()
+  list(FIND _RULE_MATRIX_KEYS "src" _SRC_KEY_INDEX)
+  if(_SRC_KEY_INDEX EQUAL -1)
+    message(SEND_ERROR "Did not find key `src` in MATRIX_KEYS: ${_RULE_MATRIX_KEYS}")
+  endif()
+
+  list(FIND _RULE_MATRIX_KEYS "target_backends" _TARGET_BACKENDS_KEY_INDEX)
+  if(_TARGET_BACKENDS_KEY_INDEX EQUAL -1)
+    message(SEND_ERROR "Did not find key `target_backends` in MATRIX_KEYS: ${_RULE_MATRIX_KEYS}")
+  endif()
+  math(EXPR _MAX_INDEX "${_MATRIX_KEYS_COUNT} - 1")
+
+  # Process failing configurations, expanding empty entries to be all entries
+  # for that key.
+  set(_FAILING_CONFIGURATIONS "")
+  foreach(_FAILING_CONFIGURATION IN LISTS _RULE_FAILING_CONFIGURATIONS)
+    string(REPLACE "," ";" _CONFIGURATION_LIST "${_FAILING_CONFIGURATION}")
+    list(LENGTH _CONFIGURATION_LIST _CONFIGURATION_KEY_COUNT)
+    if(NOT _CONFIGURATION_KEY_COUNT EQUAL _MATRIX_KEYS_COUNT)
+      message(SEND_ERROR
+        "Failing configuration ${_FAILING_CONFIGURATION} does not have same number of entries (${_MATRIX_KEY_COUNT}) as MATRIX_KEYS")
+    endif()
+
+    # message("Processing... ${_CONFIGURATION_LIST}")
```

Comment:
Want to keep these `message()` lines that are commented out, or remove them?

Comments could still be useful, since there's quite a bit of looping / conditional logic here. Maybe extracting some helper functions could make this easier to work with?

Reply:
![how_i_debug](https://user-images.githubusercontent.com/5732088/111195743-d110ac80-8579-11eb-9125-5db82bbc8f08.gif)


Yeah I can take a look at helper functions. My experience with CMake is that function calling conventions are so bad that frequently they just make things harder, but I'll take a swing at it.

Reply:
I tried and failed to extract logic into CMake functions, but I did add some comments.